Add initial version of the realm metadata tool

p-sawicki2 opened the pull request.

p-sawicki2 committed:
Add initial version of the tool

Signed-off-by: Piotr Sawicki <p.sawicki2@samsung.com>

p-sawicki2 committed:
Add unit test

Also the code has been slightly refactored to
move all crypto specific things into crypto.rs

Signed-off-by: Piotr Sawicki <p.sawicki2@samsung.com>

L0czek commented on src/utils.rs:
```diff
@@ -0,0 +1,40 @@
+    Ok(decoded)
+}
+
+pub fn arr_to_hex(bytes: &[u8]) -> String {
+    bytes.iter().fold(String::new(), |mut output, b| {
+        let _ = write!(output, "{b:02X}");
+        output
+    })
+}
```
The hex crate you include in this project has an encode function which does exactly what the arr_to_hex functions intends to do.

p-sawicki2 replied:
done

L0czek commented on src/utils.rs:
```diff
@@ -0,0 +1,40 @@
+    })
+}
+
+pub fn arr_to_string(arr: &[u8]) -> String {
+    let len = arr.iter().position(|&x| x == 0).unwrap_or(arr.len());
+    let valid_bytes = &arr[..len];
+    String::from_utf8(valid_bytes.to_vec()).expect("utf8 conversion error")
```
Do you intend to convert a c-string (null terminated array) to a rust string? If so take a look at this https://doc.rust-lang.org/std/ffi/struct.CString.html

p-sawicki2 replied:
The string might be not null-terminated (the case where the whole slice is filled up with characters). Also, the CString require unsafe code blocks, so I would prefer to leave this function as it is.

L0czek commented on src/cli.rs:
How do you feel about adding default values to the parameters that align with the instruction specified in the README?

p-sawicki2 replied:
done

L0czek reviewed (approved):
It appears to be working correctly as advertised in the README.

p-sawicki2 committed:
Apply changes according to the remarks from the review

Use hex crate to dump hexadecimal content of arrays.
Add default command line options aligned with README.md.

Signed-off-by: Piotr Sawicki <p.sawicki2@samsung.com>
